### ISSUE TITLE
python3-lxc on PyPi

### ISSUE THREAD

**Issue by run2fail** [lxc/lxc#392]
Hi,

as of today the lxc Python module that is part of the official lxc source tree is not yet available via PyPi. Instead there are multiple alternative modules that try to do (nearly) the same, for example:
- [pylxc](https://github.com/cloud9ers/pylxc)
- [pylxclite](https://github.com/Unixlike/pylxclite)
- [lxc4u](https://github.com/ravenac95/lxc4u)

It would be great if the lxc project team would consider an upload to PyPi because...
- you cannot specify the official lxc module as dependency in `setup.py` or `requirements.txt` for automatic installation
- this makes bulding/testing on CI systems problematic as the lxc module cannot be installed via the normal processes (and a distribution specific package may not be available/installable)
- it would make the official module better known and avoid that developers "reinvent the wheel". I only learned about the lxc module's existence via [Stéphane's blog posts](https://www.stgraber.org/2012/09/28/introducing-the-python-lxc-api/)

Kind regards,

BB

**Comment by stgraber** [comment]
Is maintaining python3-lxc (and possibly python2-lxc) on pypi something you'd be interested in?

Also, can pypi support multiple major versions in parallel? So we can have both 1.1 and 1.0 in there? (otherwise we should only publish 1.0).

**Comment by run2fail** [comment]
It would certainly interest me to support the lxc project by bringing ``python*-lxc`` to pypi. Unfortunately, I currently lack experience because I was planning to bring my first Python application ([Locker](https://github.com/run2fail/locker) a fig-like frontend for lxc containers based on your excellent bindings) to pypi once all dependencies can be downloaded via ``pip``.

Take the following info with a grain of salt due to my lack of experience:
* You can provide different versions via pypi, i.e., the older versions remain available. This is sometimes even required, if a particular Python application specifies an explicit version dependency in ``setup.py``.
* Both versions (for Python 2 and 3) can be managed separately, if their name differs - at least that seems to be the case with a other packages like [protobu](https://pypi.python.org/pypi?%3Aaction=search&term=protobuf&submit=search).
* In general I prefer distribution provided packages as probably anyone can upload anything to pypi. Unfortunately, many available services on the web support only installation of Python dependencies via ``pip``. For example, travis-ci prevents the use of Python packages installed via Debian packages inside ``virtualenv`` environments.

**Comment by tebanep** [comment]
I'm also creating a wrapper on top of your python binding called "eclxc" (i.e. Easy Lex-cee) which aims to bring Dockerfiles compatibility with lxc and used them as an alternative to bash templates (which I find very cumbersome to use). Right now I'm having run2fail's same problem: "pip install lxc" just doesn't work because the package is not registered in PyPI. Testing my program is quite difficult right now due to its dependence on my container's system-wide python3  (yes, i'm doing all my testing creating lxc containers inside a lxc container as instructed in your awesome blog @stgraber ... someone recalls inception?). So, I'm also eagerly waiting for lxc.py to be in PyPI. Stéphane thanks for your great contributions to the lxc project, tried Docker for a few months but found it pretty unstable and really never realized what was happening behind the scenes... with lxc everything is much more predictable and that brings me more confidence, especially at storing data and sharing it with the host and other containers (something that in Docker is quite complicated resorting to a so called "data volume container"). I find much more potential in lxc and I thing just some more things are lacking in its ecosystem, such as a full-featured gui which I also want to make using django + angularjs (the lxc web panel is pretty good, but lacks some functionality and testing) and a place to share container configuration like Docker hub. In this last topic, Flockport.com looks promising but again, the use of bash templates to create new containers its an obstacle and I think there's a need for something easier  (if Dockerfiles parsing becomes unsuitable then maybe a DSL based on yaml can make it). Again, thank you for your hard work @stgraber, I  hope this issue can be solved in the short term.

**Comment by amezin** [comment]
Anybody working on it?

Before being published on pypi, Python bindings should become 'self-contained', so that sdist packages could be built with only lxc-dev package installed (not the entire source tree).

Currently I see two problems:
- setup.py is 'configured' by autotools, and contains hardcoded paths. These paths could be passed to setup.py as command-line parameters, or as environment variables.
- lxc.c depends on private headers, not available from lxc-dev package. However, there are out-of-tree bindings for Python 2, and the problem is solved there somehow (simply copy-paste all necessary functions?)

Are there any other obstacles I didn't notice?

I'm interested in getting this done, and it doesn't seem like a big amount of work. However, I'd like to see some feedback before I start working on it, especially regarding copy-pasting from private headers.

@stgraber ?

**Comment by stgraber** [comment]
I'm not necessarily opposed to having python3-lxc buildable out of tree, but if that's to be done, its dependency on private headers will have to be fixed the right way.

That is, the resulting .so can't call any private symbol as we don't make any kind of guarantee for things outside of our public API.

**Comment by amezin** [comment]
TODOs, mostly for myself, but any comments are welcome too:
- "_lxc" is a bad package name, because it starts with an underscore. It can't be specified as a requirement in setup.py. Why does the package (not the extension) have such obscure name?
- Source distributions (setup.py sdist) don't include 'examples' folder currently.
- The package has almost no metadata.
- "Standalone" python-lxc builds aren't tested on Travis-ci yet.

By the way, @stgraber what do you think about moving python-lxc to its own separate repository? If it will be published on PyPI, separate repo will be much more convenient. At least for me as a user. Also, distutils and autotools don't integrate very well with each other...

**Comment by amezin** [comment]
python-lxc now can be installed with pip in a virtualenv:
$ pip install git+ssh://git@github.com/lxc/lxc.git#subdirectory=src/python-lxc

**Comment by stgraber** [comment]
@amezin So I'd rather not split it out of the main repository. Moving it out would mean more effort on our end when doing releases, would require its own testing infrastructure and would also double the packaging effort for Linux distributions.

**Comment by brauner** [comment]
There now is a python3-lxc repo and the bindings will soon be removed from the lxc repo completely.
